// oceanic.js/dist/lib/structures/StageInstance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
/** @module StageInstance */
const Base_1 = tslib_1.__importDefault(require("./Base"));
/** Represents a stage instance. */
class StageInstance extends Base_1.default {
    _cachedChannel;
    _cachedGuild;
    _cachedScheduledEvent;
    /** The ID of the associated stage channel. */
    channelID;
    /** @deprecated If the stage channel is discoverable */
    discoverableDisabled;
    /** The id of the guild associated with this stage instance's stage channel. */
    guildID;
    /** The [privacy level](https://discord.com/developers/docs/resources/stage-instance#stage-instance-object-privacy-level) of this stage instance. */
    privacyLevel;
    /** The id of the scheduled event for this stage instance, if applicable. */
    scheduledEventID;
    /** The topic of this stage instance. */
    topic;
    constructor(data, client) {
        super(data.id, client);
        this.channelID = data.channel_id;
        this.discoverableDisabled = !!data.discoverable_disabled;
        this.guildID = data.guild_id;
        this.privacyLevel = data.privacy_level;
        this.scheduledEventID = data.guild_scheduled_event_id;
        this.topic = data.topic;
        this.update(data);
    }
    update(data) {
        if (data.channel_id !== undefined) {
            this.channelID = data.channel_id;
        }
        if (data.discoverable_disabled !== undefined) {
            this.discoverableDisabled = data.discoverable_disabled;
        }
        if (data.guild_scheduled_event_id !== undefined) {
            this.scheduledEventID = data.guild_scheduled_event_id;
        }
        if (data.privacy_level !== undefined) {
            this.privacyLevel = data.privacy_level;
        }
        if (data.topic !== undefined) {
            this.topic = data.topic;
        }
    }
    /** The associated stage channel. */
    get channel() {
        return this._cachedChannel ?? (this._cachedChannel = this.client.getChannel(this.channelID));
    }
    /** The guild of the associated stage channel. This will throw an error if the guild is not cached. */
    get guild() {
        if (!this._cachedGuild) {
            this._cachedGuild = this.client.guilds.get(this.guildID);
            if (!this._cachedGuild) {
                throw new Error(`${this.constructor.name}#guild is not present if you don't have the GUILDS intent.`);
            }
        }
        return this._cachedGuild;
    }
    /** The scheduled event for this stage instance, if applicable. */
    get scheduledEvent() {
        if (this.scheduledEventID !== null && this._cachedScheduledEvent !== null) {
            try {
                return this._cachedScheduledEvent ?? (this._cachedScheduledEvent = this.guild.scheduledEvents.get(this.scheduledEventID));
            }
            catch {
                return (this._cachedScheduledEvent = undefined);
            }
        }
        return this._cachedScheduledEvent === null ? this._cachedScheduledEvent : (this._cachedScheduledEvent = null);
    }
    toJSON() {
        return {
            ...super.toJSON(),
            channelID: this.channelID,
            discoverableDisabled: this.discoverableDisabled,
            guildID: this.guildID,
            privacyLevel: this.privacyLevel,
            scheduledEventID: this.scheduledEventID,
            topic: this.topic
        };
    }
}
exports.default = StageInstance;
