// @oceanicjs/builders/dist/Structures/Button.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Component_1 = __importDefault(require("./Component"));
const oceanic_js_1 = require("oceanic.js");
class Button extends Component_1.default {
    customID;
    emoji;
    label;
    style;
    url;
    /**
     * Create a new Button.
     * @param style The style of this button.
     * @param urlOrCustomID The url of this button if style 5, else the custom ID of this button.
     */
    constructor(style, urlOrCustomID) {
        super(oceanic_js_1.ComponentTypes.BUTTON);
        this.style = style;
        if (style === oceanic_js_1.ButtonStyles.LINK) {
            this.url = urlOrCustomID;
        }
        else {
            this.customID = urlOrCustomID;
        }
    }
    /** this method is meant to be for internal use only, don't use it, as it may break or change at a moments notice */
    load(style, urlOrCustomID, label, emoji, disabled) {
        if (style) {
            this.setStyle(style);
        }
        if (urlOrCustomID) {
            if (this.style === oceanic_js_1.ButtonStyles.LINK) {
                this.url = urlOrCustomID;
            }
            else {
                this.customID = urlOrCustomID;
            }
        }
        if (label) {
            this.setLabel(label);
        }
        if (emoji) {
            this.setEmoji(emoji);
        }
        if (typeof disabled !== "undefined") {
            if (disabled) {
                this.disable();
            }
            else {
                this.enable();
            }
        }
        return this;
    }
    /**
     * Set the custom id of this button (styles 1-4).
     * @param customID A developer-defined identifier for the button, max 100 characters.
     */
    setCustomID(customID) {
        this.customID = customID;
        return this;
    }
    /**
     * Set the emoji of this button.
     * @param emoji The emoji to display on this button.
     */
    setEmoji(emoji) {
        this.emoji = emoji;
        return this;
    }
    /**
     * Set the label of this button
     * @param label The label to display on this button.
     */
    setLabel(label) {
        this.label = label;
        return this;
    }
    /**
     * Set the style of this button.
     * * 1 - blurple
     * * 2 - grey
     * * 3 - green
     * * 4 - red
     * * 5 - link
     * @param style The [style](https://discord.com/developers/docs/interactions/message-components#button-object-button-styles) of this button.
     */
    setStyle(style) {
        this.style = style;
        return this;
    }
    /**
     * Set the custom id of this button (style 5).
     * @param url The url to open when this button is clicked.
     */
    setURL(url) {
        this.url = url;
        return this;
    }
    toJSON() {
        const obj = {
            type: this.type,
            style: this.style,
            label: this.label,
            emoji: this.emoji,
            disabled: this.disabled
        };
        if (this.style === oceanic_js_1.ButtonStyles.LINK) {
            obj.url = this.url;
            return obj;
        }
        else {
            obj.customID = this.customID;
            return obj;
        }
    }
    toJSONRaw() {
        const json = this.toJSON();
        if (this.style === oceanic_js_1.ButtonStyles.LINK) {
            return json;
        }
        else {
            const raw = json;
            raw.custom_id = raw.customID;
            delete raw.customID;
            return raw;
        }
    }
}
exports.default = Button;
