// @oceanicjs/builders/dist/Structures/SelectMenu.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Component_1 = __importDefault(require("./Component"));
class SelectMenu extends Component_1.default {
    channelTypes;
    customID;
    disabled = false;
    maxValues;
    minValues;
    options = [];
    placeholder;
    /**
     * Create a new SelectMenu.
     * @param customID The custom ID of this select menu.
     */
    constructor(type, customID) {
        super(type);
        this.customID = customID;
    }
    /** this method is meant to be for internal use only, don't use it, as it may break or change at a moments notice */
    load(customID, options, placeholder, minValues, maxValues, disabled) {
        if (customID) {
            this.setCustomID(customID);
        }
        if (options && Array.isArray(options) && options.length > 0) {
            this.clearOptions();
            this.addOptions(...options);
        }
        if (placeholder) {
            this.setPlaceholder(placeholder);
        }
        if (minValues) {
            this.setValues(minValues, undefined);
        }
        if (maxValues) {
            this.setValues(undefined, maxValues);
        }
        if (typeof disabled !== "undefined") {
            if (disabled) {
                this.disable();
            }
            else {
                this.enable();
            }
        }
        return this;
    }
    /**
     * Add an option to this select menu.
     * @param label The label for this option.
     * @param value The value of this option.
     * @param description The description of this option.
     * @param emoji The emoji to displayed with this option.
     */
    addOption(label, value, description, emoji, defaultSelection) {
        this.options.push({
            label,
            value,
            description,
            emoji,
            default: defaultSelection
        });
        return this;
    }
    /**
     * Add options to this select menu in bulk.
     * @param options The options to add.
     */
    addOptions(...options) {
        options.map(o => this.addOption(o.label, o.value, o.description, o.emoji, o.default));
        return this;
    }
    /** Clear all currently present options on this select menu. */
    clearOptions() {
        this.options = [];
        return this;
    }
    /** Set the valid channel types for a channel select menu. */
    setChannelTypes(...types) {
        if (types.length === 1 && Array.isArray(types[0])) {
            types = types[0];
        }
        this.channelTypes = types;
        return this;
    }
    /**
     * Set the custom ID of this select menu.
     * @param customID A developer-defined identifier for the button, max 100 characters.
     */
    setCustomID(customID) {
        this.customID = customID;
        return this;
    }
    /**
     * Set the placeholder of this select menu.
     * @param placeholder Custom placeholder text if nothing is selected, max 100 characters.
     */
    setPlaceholder(placeholder) {
        this.placeholder = placeholder;
        return this;
    }
    /**
     * Set the minimum/maximum values of this select menu.
     * @param min The minimum selected values.
     * @param max The maximum selected values.
     */
    setValues(min, max) {
        if (min) {
            this.minValues = min;
        }
        if (max) {
            this.maxValues = max;
        }
        return this;
    }
    /** converts this SelectMenu instance to json. */
    toJSON() {
        return {
            type: this.type,
            customID: this.customID,
            options: this.options,
            placeholder: this.placeholder,
            minValues: this.minValues,
            maxValues: this.maxValues,
            disabled: this.disabled,
            channelTypes: this.channelTypes
        };
    }
    /** converts this SelectMenu instance to json. */
    toJSONRaw() {
        return {
            type: this.type,
            custom_id: this.customID,
            options: this.options,
            placeholder: this.placeholder,
            min_values: this.minValues,
            max_values: this.maxValues,
            disabled: this.disabled,
            channel_types: this.channelTypes
        };
    }
}
exports.default = SelectMenu;
