// @oceanicjs/builders/dist/ApplicationCommandOptionBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const oceanic_js_1 = require("oceanic.js");
class ApplicationCommandOptionBuilder {
    autocomplete;
    channelTypes;
    choices = [];
    description = "";
    descriptionLocalizations;
    max;
    min;
    name;
    nameLocalizations;
    options = [];
    parent;
    required;
    type;
    constructor(type, name) {
        this.type = type;
        this.name = name;
    }
    /**
     * Add a choice.
     * @param name The name of the choice.
     * @param value The value of the choice.
     * @param nameLocalizations A map of [locales](https://discord.com/developers/docs/reference#locales) to name localizations.
     */
    addChoice(name, value, nameLocalizations) {
        this.choices.push({ name, value, nameLocalizations });
        return this;
    }
    /**
     * Add a description localization.
     * @param locale The [locale](https://discord.com/developers/docs/reference#locales) of the localization.
     * @param description The localized description.
     */
    addDescriptionLocalization(locale, description) {
        (this.descriptionLocalizations ??= {})[locale] = description;
        return this;
    }
    /**
     * Add a name localization.
     * @param locale The [locale](https://discord.com/developers/docs/reference#locales) of the localization.
     * @param name The localized name.
     */
    addNameLocalization(locale, name) {
        (this.nameLocalizations ??= {})[locale] = name;
        return this;
    }
    addOption(...args) {
        if (args.length === 1) {
            if (args[0] instanceof ApplicationCommandOptionBuilder) {
                args[0] = args[0].toJSON();
            }
            this.options.push(args[0]);
            return this;
        }
        else {
            const [name, type, extra] = args;
            const option = new ApplicationCommandOptionBuilder(type, name);
            if (extra) {
                if (typeof extra === "function") {
                    extra.call(this, option);
                    this.options.push(option.toJSON());
                }
                else {
                    this.options.push({ ...extra, name, type });
                }
            }
            else {
                this.options.push(option.toJSON());
            }
            return this;
        }
    }
    /** Toggle autocomplete for this option. */
    setAutocomplete(value = !this.autocomplete) {
        this.autocomplete = value;
        return this;
    }
    /** Set the allowed channel types for this option. */
    setChannelTypes(types) {
        this.channelTypes = types;
        return this;
    }
    /** Set the choices for this option. */
    setChoices(choices) {
        this.choices = choices;
        return this;
    }
    /** Set the description of this option. */
    setDescription(description) {
        this.description = description;
        return this;
    }
    /**
     * Set the description localizations of this option.
     * @param localizations A map of [locales](https://discord.com/developers/docs/reference#locales) to localized description strings.
     */
    setDescriptionLocalizations(localizations) {
        this.descriptionLocalizations = localizations;
        return this;
    }
    /** Set the min/max for this option. This applied to both `minValue`/`maxValue` and `minLength`/`maxLength`. */
    setMinMax(min, max) {
        this.max = max;
        this.min = min;
        return this;
    }
    /** Set the name of this option. */
    setName(name) {
        this.name = name;
        return this;
    }
    /**
     * Set the name localizations of this option.
     * @param localizations A map of [locales](https://discord.com/developers/docs/reference#locales) to localized name strings.
     */
    setNameLocalizations(localizations) {
        this.nameLocalizations = localizations;
        return this;
    }
    /** Toggle this option being required. */
    setRequired(value = !this.required) {
        this.required = value;
        return this;
    }
    /** Convert this command to JSON. */
    toJSON() {
        let res;
        switch (this.type) {
            case oceanic_js_1.ApplicationCommandOptionTypes.SUB_COMMAND: {
                res = {
                    type: this.type,
                    name: this.name,
                    description: this.description,
                    options: this.options
                };
                break;
            }
            case oceanic_js_1.ApplicationCommandOptionTypes.SUB_COMMAND_GROUP: {
                res = {
                    type: this.type,
                    name: this.name,
                    description: this.description,
                    options: this.options
                };
                break;
            }
            case oceanic_js_1.ApplicationCommandOptionTypes.STRING: {
                res = {
                    type: this.type,
                    name: this.name,
                    description: this.description,
                    autocomplete: this.autocomplete,
                    choices: this.choices,
                    minLength: this.min,
                    maxLength: this.max,
                    required: this.required
                };
                break;
            }
            case oceanic_js_1.ApplicationCommandOptionTypes.INTEGER:
            case oceanic_js_1.ApplicationCommandOptionTypes.NUMBER: {
                res = {
                    type: this.type,
                    name: this.name,
                    description: this.description,
                    choices: this.choices,
                    minValue: this.min,
                    maxValue: this.max,
                    required: this.required
                };
                break;
            }
            case oceanic_js_1.ApplicationCommandOptionTypes.BOOLEAN:
            case oceanic_js_1.ApplicationCommandOptionTypes.USER:
            case oceanic_js_1.ApplicationCommandOptionTypes.ROLE:
            case oceanic_js_1.ApplicationCommandOptionTypes.MENTIONABLE:
            case oceanic_js_1.ApplicationCommandOptionTypes.ATTACHMENT: {
                res = {
                    type: this.type,
                    name: this.name,
                    description: this.description,
                    required: this.required
                };
                break;
            }
            case oceanic_js_1.ApplicationCommandOptionTypes.CHANNEL: {
                res = {
                    type: this.type,
                    name: this.name,
                    description: this.description,
                    channelTypes: this.channelTypes,
                    required: this.required
                };
                break;
            }
            default: {
                return null;
            }
        }
        return res;
    }
}
exports.default = ApplicationCommandOptionBuilder;
