// @oceanicjs/builders/dist/ApplicationCommandBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ApplicationCommandOptionBuilder_1 = __importDefault(require("./ApplicationCommandOptionBuilder"));
const oceanic_js_1 = require("oceanic.js");
const permissionNames = Object.keys(oceanic_js_1.Permissions).filter(p => !/^[0-9]+$/.test(p));
class ApplicationCommandBuilder {
    defaultMemberPermissions;
    description;
    descriptionLocalizations;
    dmPermission;
    name;
    nameLocalizations;
    options = [];
    type;
    constructor(type, name) {
        this.type = type;
        this.name = name;
    }
    /**
     * Add a description localization.
     * @param locale The [locale](https://discord.com/developers/docs/reference#locales) of the localization.
     * @param description The localized description.
     */
    addDescriptionLocalization(locale, description) {
        (this.descriptionLocalizations ??= {})[locale] = description;
        return this;
    }
    /**
     * Add a name localization.
     * @param locale The [locale](https://discord.com/developers/docs/reference#locales) of the localization.
     * @param name The localized name.
     */
    addNameLocalization(locale, name) {
        (this.nameLocalizations ??= {})[locale] = name;
        return this;
    }
    addOption(...args) {
        if (args.length === 1) {
            if (args[0] instanceof ApplicationCommandOptionBuilder_1.default) {
                args[0] = args[0].toJSON();
            }
            this.options.push(args[0]);
            return this;
        }
        else {
            const [name, type, extra] = args;
            const option = new ApplicationCommandOptionBuilder_1.default(type, name);
            if (extra) {
                if (typeof extra === "function") {
                    extra.call(this, option);
                    this.options.push(option.toJSON());
                }
                else {
                    this.options.push({ ...extra, name, type });
                }
            }
            else {
                this.options.push(option.toJSON());
            }
            return this;
        }
    }
    /** Allow this command to be used in direct messages. */
    allowDMUsage() {
        this.setDMPermission(true);
        return this;
    }
    /** Disallow this command from being used in direct messages. */
    disallowDMUsage() {
        this.setDMPermission(false);
        return this;
    }
    /** Set if this command can be used in direct messages. */
    setDMPermission(dmPermission) {
        this.dmPermission = dmPermission;
        return this;
    }
    /** Set the default permissions required to use this command. */
    setDefaultMemberPermissions(...permissions) {
        let data;
        if (permissions.length > 1 || (typeof permissions[0] === "string" && permissionNames.includes(permissions[0]))) {
            data = permissions;
        }
        else {
            data = permissions[0];
        }
        if (data instanceof oceanic_js_1.Permission) {
            data = data.allow.toString();
        }
        else if (typeof data === "bigint") {
            data = data.toString();
        }
        else if (Array.isArray(data)) {
            data = data.map(perm => oceanic_js_1.Permissions[perm]).reduce((a, b) => a | b, 0n).toString();
        }
        this.defaultMemberPermissions = data;
        return this;
    }
    /** Set the description of this command. */
    setDescription(description) {
        this.description = description;
        return this;
    }
    /**
     * Set the description localizations for this command.
     * @param localizations A map of [locales](https://discord.com/developers/docs/reference#locales) to localized description strings.
     */
    setDescriptionLocalizations(localizations) {
        this.descriptionLocalizations = localizations;
        return this;
    }
    /** Set the name of this command.*/
    setName(name) {
        this.name = name;
        return this;
    }
    /**
     * Set the name localizations for this command.
     * @param localizations A map of [locales](https://discord.com/developers/docs/reference#locales) to localized name strings.
     */
    setNameLocalizations(localizations) {
        this.nameLocalizations = localizations;
        return this;
    }
    /** Convert this command to JSON. */
    toJSON() {
        return {
            defaultMemberPermissions: this.defaultMemberPermissions,
            description: this.description,
            descriptionLocalizations: this.descriptionLocalizations,
            dmPermission: this.dmPermission,
            name: this.name,
            nameLocalizations: this.nameLocalizations,
            options: this.options,
            type: this.type
        };
    }
}
exports.default = ApplicationCommandBuilder;
