// @oceanicjs/builders/dist/ComponentBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ActionRow_1 = __importDefault(require("./Structures/ActionRow"));
const Button_1 = __importDefault(require("./Structures/Button"));
const SelectMenu_1 = __importDefault(require("./Structures/SelectMenu"));
const TextInput_1 = __importDefault(require("./Structures/TextInput"));
const oceanic_js_1 = require("oceanic.js");
const SelectMenuTypeValues = [oceanic_js_1.ComponentTypes.STRING_SELECT, oceanic_js_1.ComponentTypes.USER_SELECT, oceanic_js_1.ComponentTypes.ROLE_SELECT, oceanic_js_1.ComponentTypes.MENTIONABLE_SELECT, oceanic_js_1.ComponentTypes.CHANNEL_SELECT];
const UniqueRowNeeded = [...SelectMenuTypeValues, oceanic_js_1.ComponentTypes.TEXT_INPUT];
class ComponentBuilder {
    currentIndex = 0;
    rows = [];
    rowMax = 5;
    constructor(rowMax) {
        if (rowMax) {
            this.rowMax = rowMax;
        }
    }
    /**
     * Convert an emoji to a partial.
     * @param emoji The unicode point of the emoji if default, else the fully qualified emoji.
     * @param type `default` if built in (unicode), `custom` otherwise (this can be assumed from the format by us).
     * @example emojiToPartial("🐾", "default")
     * @example emojiToPartial("<:paws8:681748079778463796>", "custom")
     */
    static emojiToPartial(emoji, type = "custom") {
        if (type === "default") {
            return {
                id: null,
                name: emoji,
                animated: false
            };
        }
        else {
            const [, anim, name, id] = /^<?(a)?:(.*):([\d]{15,21})>?$/.exec(emoji) ?? [];
            if (!name || !id) {
                return this.emojiToPartial(emoji, "default");
            }
            return {
                id,
                name,
                animated: anim === "a"
            };
        }
    }
    getCurrentRow() {
        return (this.rows[this.currentIndex] || (this.rows[this.currentIndex] = new ActionRow_1.default()));
    }
    /**
     * Add a component to the current row, or a new row depending on certain conditions
     * @param component The component to add.
     */
    addComponent(component) {
        const cur = this.getCurrentRow();
        if (UniqueRowNeeded.includes(component.type)) {
            if (cur.isEmpty()) {
                cur.addComponent(component);
                this.addRow();
            }
            else {
                this.addRow([component]).addRow();
            }
        }
        else {
            if (cur.size >= this.rowMax) {
                this.addRow([component]);
            }
            else {
                cur.addComponent(component);
            }
        }
        return this;
    }
    /**
     * Add several components.
     * @param components The components to add.
     */
    addComponents(...components) {
        components.map(c => this.addComponent(c));
        return this;
    }
    /**
     * Add an interaction button to the current row.
     * @param options The options for adding the interaction button.
     */
    addInteractionButton(options) {
        this.addComponent(new Button_1.default(options.style, options.customID)["load"](options.style, options.customID, options.label, options.emoji, options.disabled));
        return this;
    }
    /**
     * Start a new action row.
     * @param components The components to start this new row with.
     */
    addRow(components = []) {
        this.currentIndex++;
        this.rows.push(new ActionRow_1.default().addComponents(...components));
        return this;
    }
    /**
     * Add a select menu (to the current row, if empty - else as a new row).
     * @param options The options for adding the select menu.
     */
    addSelectMenu(options) {
        this.addComponent(new SelectMenu_1.default(options.type, options.customID)["load"](options.customID, options.options, options.placeholder, options.minValues, options.maxValues, options.disabled));
        return this;
    }
    /**
     * Add a text input to the current row.
     * @param options The options for adding the text input.
     * @returns
     */
    addTextInput(options) {
        this.addComponent(new TextInput_1.default(options.style, options.label, options.customID)["load"](options.style, options.label, options.customID, options.placeholder, options.value, options.minLength, options.maxLength, options.required));
        return this;
    }
    /**
     * Add a url button to the current row
     * @param options The options for adding the url button.
     */
    addURLButton(options) {
        this.addComponent(new Button_1.default(oceanic_js_1.ButtonStyles.LINK, options.url)["load"](oceanic_js_1.ButtonStyles.LINK, options.url, options.label, options.emoji, options.disabled));
        return this;
    }
    /**
     * Remove all of the rows that are empty.
     */
    removeEmptyRows() {
        this.rows.forEach((row, index) => {
            if (row.size === 0) {
                this.rows.splice(index, 1);
            }
        });
        this.currentIndex = this.rows.length - 1;
        return this;
    }
    setRowMax(rowMax) {
        this.rowMax = rowMax;
        return this;
    }
    /** convert the current contents to JSON */
    toJSON() {
        return this.removeEmptyRows().rows.map(row => row.toJSON());
    }
    /** convert the current contents to JSON */
    toJSONRaw() {
        return this.removeEmptyRows().rows.map(row => row.toJSONRaw());
    }
}
exports.default = ComponentBuilder;
