// @oceanicjs/builders/dist/EmbedBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const oceanic_js_1 = require("oceanic.js");
class EmbedBuilder {
    json = {};
    static loadFromJSON(json, forceSingular) {
        if (Array.isArray(json)) {
            const val = json.map(v => EmbedBuilder.loadFromJSON(v));
            return forceSingular ? val[0] : val;
        }
        else {
            return new EmbedBuilder().load(json);
        }
    }
    /**
     * load json into this embed builder instance - use static loadFromJSON method
     * @private
     * @param {EmbedOptions} json - the json to load
     * @returns {this}
     */
    load(json) {
        this.json = json;
        return this;
    }
    /**
     * Add a blank field to the embed (zero width spaces).
     * @param inline If the field should be displayed inline.
     */
    addBlankField(inline) {
        return this.addField("\u200b", "\u200b", inline);
    }
    /**
     * Add a field to the embed.
     * @param name The field name.
     * @param value The field value.
     * @param inline If the field should be inline.
     */
    addField(name, value, inline) {
        this.json.fields = [...(this.json.fields ?? []), { name, value, inline }];
        return this;
    }
    /**
     * Add multiple fields.
     * @param fields - the fields to add
     */
    addFields(...fields) {
        fields.forEach(arg => this.addField(arg.name, arg.value, arg.inline));
        return this;
    }
    /**
     * Get the current author.
     */
    getAuthor() {
        return this.json.author;
    }
    /**
     * Get the current color.
     */
    getColor() {
        return this.json.color;
    }
    /**
     * Get the current description.
     */
    getDescription() {
        return this.json.description;
    }
    /**
     * Get the field at the specified index.
     * @param index The index of the field to get.
     */
    getField(index) {
        return (this.json.fields ?? [])[index];
    }
    /**
     * Get the current fields.
     */
    getFields() {
        return (this.json.fields ?? []);
    }
    /**
     * Get the current footer.
     */
    getFooter() {
        return this.json.footer;
    }
    /**
     * Get the current image.
     */
    getImage() {
        return this.json.image;
    }
    /**
     * Get the current thumbnail.
     */
    getThumbnail() {
        return this.json.thumbnail;
    }
    /**
     * Get the current timestamp.
     */
    getTimestamp() {
        return this.json.timestamp;
    }
    /**
     * Get the current timestamp as a date instance.
     */
    getTimestampDate() {
        return !this.json.timestamp ? undefined : new Date(this.json.timestamp);
    }
    /**
     * Get the current title.
     */
    getTitle() {
        return this.json.title;
    }
    /**
     * Get the current url.
     */
    getURL() {
        return this.json.url;
    }
    /**
     * remove the current author
     * @returns {this}
     */
    removeAuthor() {
        this.json.author = undefined;
        return this;
    }
    /**
     * Remove the current color.
     */
    removeColor() {
        this.json.color = undefined;
        return this;
    }
    /**
     * Remove the current description.
     */
    removeDescription() {
        this.json.description = undefined;
        return this;
    }
    /**
     * Remove the current footer.
     */
    removeFooter() {
        this.json.footer = undefined;
        return this;
    }
    /**
     * Remove the current image.
     */
    removeImage() {
        this.json.image = undefined;
        return this;
    }
    /**
     * Remove the current thumbnail.
     */
    removeThumbnail() {
        this.json.thumbnail = undefined;
        return this;
    }
    /**
     * Remove the current timestamp.
     */
    removeTimestamp() {
        this.json.timestamp = undefined;
        return this;
    }
    /**
     * Remove the current title.
     */
    removeTitle() {
        this.json.title = undefined;
        return this;
    }
    /**
     * Remove the current url.
     */
    removeURL() {
        this.json.url = undefined;
        return this;
    }
    /**
     * set the embed author
     * @param name The name of the author.
     * @param iconURL An icon url for the author.
     * @param url A url for the author.
     */
    setAuthor(name, iconURL, url) {
        this.json.author = {
            name,
            iconURL,
            url
        };
        return this;
    }
    /**
     * Set the embed color.
     * @param color The color.
     */
    setColor(color) {
        this.json.color = color;
        return this;
    }
    /**
     * Set the embed description.
     * @param value The description. A string, array of strings, or both spread across multiple parameters. They will be joined by LF charactes.
     */
    setDescription(first, ...other) {
        this.json.description = [...(Array.isArray(first) ? first : [first]), ...(other.map(o => [...(Array.isArray(o) ? o : [o])].join("\n")))].join("\n");
        return this;
    }
    /**
     * Set the embed footer.
     * @param text - The text.
     * @param iconURL - The icon url.
     */
    setFooter(text, iconURL) {
        this.json.footer = { text, iconURL };
        return this;
    }
    /**
     * Set the embed image.
     * @param url The Image url.
     */
    setImage(url) {
        this.json.image = { url };
        return this;
    }
    /**
     * Set the embed thumbnail.
     * @param url The thumbnail url.
     */
    setThumbnail(url) {
        this.json.thumbnail = { url };
        return this;
    }
    /**
     * Set the embed timestamp.
     * @param time An ISO 8601 timestamp, Date object, or "now".
     */
    setTimestamp(time) {
        if (time === "now") {
            time = new Date().toISOString();
        }
        else if (time instanceof Date) {
            time = time.toISOString();
        }
        this.json.timestamp = time;
        return this;
    }
    /**
     * Set the embed title.
     * @param title The title.
     */
    setTitle(title) {
        this.json.title = title;
        return this;
    }
    /**
     * Set the embed url.
     * @param url The url.
     */
    setURL(url) {
        this.json.url = url;
        return this;
    }
    toJSON(array = false) {
        return array ? [this.json] : this.json;
    }
    toJSONRaw(array = false) {
        const [embed] = oceanic_js_1.Util.prototype.embedsToRaw([this.json]);
        return array ? [embed] : embed;
    }
}
exports.default = EmbedBuilder;
