// oceanic.js/dist/lib/routes/Users.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Routes = tslib_1.__importStar(require("../util/Routes"));
const ExtendedUser_1 = tslib_1.__importDefault(require("../structures/ExtendedUser"));
/** Various methods for interacting with users. */
class Users {
    #manager;
    constructor(manager) {
        this.#manager = manager;
    }
    /** Alias for {@link Routes/Channels#createDM | Channels#createDM}. */
    get createDM() {
        return this.#manager.channels.createDM.bind(this.#manager.channels);
    }
    /**
     * Edit the currently authenticated user.
     *
     * Note: This does not touch the client's cache in any way.
     * @param options The options to edit with.
     */
    async editSelf(options) {
        if (options.avatar) {
            options.avatar = this.#manager.client.util._convertImage(options.avatar, "avatar");
        }
        return this.#manager.authRequest({
            method: "PATCH",
            path: Routes.USER("@me"),
            json: options
        }).then(data => new ExtendedUser_1.default(data, this.#manager.client));
    }
    /**
     * Get a user.
     * @param id the ID of the user
     */
    async get(id) {
        return this.#manager.authRequest({
            method: "GET",
            path: Routes.USER(id)
        }).then(data => this.#manager.client.users.update(data));
    }
    /**
     * Get the currently authenticated user's information.
     *
     * Note: This does not touch the client's cache in any way.
     * @deprecated Moved to {@link Routes/OAuth#getCurrentUser}. This will be removed in `1.5.0`.
     */
    async getCurrentUser() {
        return this.#manager.oauth.getCurrentUser();
    }
    /**
     * Leave a guild.
     * @param id The ID of the guild to leave.
     */
    async leaveGuild(id) {
        await this.#manager.authRequest({
            method: "DELETE",
            path: Routes.OAUTH_GUILD(id)
        });
    }
}
exports.default = Users;
